cmd/xfn/spark/spark.go

```diff
@@ -0,0 +1,217 @@
+	"github.com/crossplane/crossplane/internal/oci/store/uncompressed"
+)
+
+// Error strings.
+const (
+	errReadRequest      = "cannot read request from stdin"
+	errUnmarshalRequest = "cannot unmarshal request data from stdin"
+	errNewBundleStore   = "cannot create OCI runtime bundle store"
+	errNewDigestStore   = "cannot create OCI image digest store"
+	errParseRef         = "cannot parse OCI image reference"
+	errPull             = "cannot pull OCI image"
+	errBundleFn         = "cannot create OCI runtime bundle"
+	errMkRuntimeRootdir = "cannot make OCI runtime cache"
+	errRuntime          = "OCI runtime error"
+	errCleanupBundle    = "cannot cleanup OCI runtime bundle"
+	errMarshalResponse  = "cannot marshal response data to stdout"
+	errWriteResponse    = "cannot write response data to stdout"
+	errCPULimit         = "cannot limit container CPU"
+	errMemoryLimit      = "cannot limit container memory"
+	errHostNetwork      = "cannot configure container to run in host network namespace"
+)
+
+// The path within the cache dir that the OCI runtime should use for its
+// '--root' cache.
+const ociRuntimeRoot = "runtime"
+
+// The time after which the OCI runtime will be killed if none is specified in
+// the RunFunctionRequest.
+const defaultTimeout = 25 * time.Second
+
+// Command runs a containerized Composition Function.
+type Command struct {
+	CacheDir string `short:"c" help:"Directory used for caching function images and containers." default:"/xfn"`
+	Runtime  string `help:"OCI runtime binary to invoke." default:"crun"`
+}
+
+// Run a Composition Function inside an unprivileged user namespace. Reads a
+// protocol buffer serialized RunFunctionRequest from stdin, and writes a
+// protocol buffer serialized RunFunctionResponse to stdout.
+func (c *Command) Run() error { //nolint:gocyclo // TODO(negz): Refactor some of this out into functions, add tests.
+	pb, err := io.ReadAll(os.Stdin)
+	if err != nil {
+		return errors.Wrap(err, errReadRequest)
+	}
+
+	req := &v1alpha1.RunFunctionRequest{}
+	if err := proto.Unmarshal(pb, req); err != nil {
+		return errors.Wrap(err, errUnmarshalRequest)
+	}
+
+	t := req.GetRunFunctionConfig().GetTimeout().AsDuration()
+	if t == 0 {
+		t = defaultTimeout
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), t)
+	defer cancel()
+
+	runID := uuid.NewString()
+
+	// We prefer to use an overlayfs bundler where possible. It roughly doubles
+	// the disk space per image because it caches layers as overlay compatible
+	// directories in addition to the CachingImagePuller's cache of uncompressed
+	// layer tarballs. The advantage is faster start times for containers with
+	// cached image, because it creates an overlay rootfs. The uncompressed
+	// bundler on the other hand must untar all of a containers layers to create
+	// a new rootfs each time it runs a container.
+	var s store.Bundler = uncompressed.NewBundler(c.CacheDir)
+	if overlay.Supported(c.CacheDir) {
```

Comment:
Assuming this wouldn't be changed at runtime, could this be moved to initialization somewhere and passed as a config to spark?

Reply:
I'm not sure I understand what you're suggesting.

Is the idea that you'd prefer to avoid running this check every time `xfn spark` runs? If so, I agree in principal but I'm not sure its worthwhile (yet). This check needs to be done by something running in the same "environment" as spark (i.e. in the same namespaces, with the same capabilities) - it's not something we could move up into the long-running `xfn start` process for example.

Presumably we could do something like:

* Add an `xfn spark --store-type` flag to force it to use a particular type of store (overlay or uncompressed).
* Add an `xfn spark --check-capabilities` flag, which would cause spark to report what it can do and then exit.
* Have `xfn start` run `xfn spark --check-capabilities` once at startup to determine what stores it supports, then pass the desired `--store-type` flag accordingly.

This all seems pretty complex though - I'd prefer to wait and see whether we actually need to remove this check from the function run path. I believe the cost is 5-6 mkdir calls, one mount, and an unmount each time we run `xfn spark`. Right now it doesn't seem to be adding a lot of overhead to the time to run a function:

```console
# Cold image cache
$ time _output/bin/linux_arm64/xfn run negz/xfn-nop  ~/xp/fnio.yaml 1>/dev/null
_output/bin/linux_arm64/xfn run negz/xfn-nop ~/xp/fnio.yaml > /dev/null  0.20s user 0.12s system 22% cpu 1.413 total

# Hot image cache
$ time _output/bin/linux_arm64/xfn run negz/xfn-nop  ~/xp/fnio.yaml 1>/dev/null
_output/bin/linux_arm64/xfn run negz/xfn-nop ~/xp/fnio.yaml > /dev/null  0.10s user 0.03s system 107% cpu 0.117 total
```